glusterd2/pmap/registry.go

```diff
@@ -0,0 +1,207 @@
+	SignedIn bool `json:"signed-in"`
+}
+
+type registryType struct {
+	sync.RWMutex
+	basePort  int
+	lastAlloc int
+	ports     [gfPortMax + 1]portStatus
+	bricks    map[string]*brickStatus
+
+	portLockFds map[int]int
+}
+
+func (r *registryType) init() {
+	for i := r.basePort; i <= gfPortMax; i++ {
+		// TODO: There is a race in this check when there are multiple
+		// glusterd2 instances running on the same machine.
+		if isPortFree(i) {
+			r.ports[i].State = GfPmapPortFree
+		} else {
+			r.ports[i].State = GfPmapPortForeign
+		}
+		r.ports[i].Port = i
+	}
+}
+
+// Allocate finds a free port and returns the port number. The
+// allocated free port will be in leased state until the brick
+// process it was leased to sends a SIGN IN request to glusterd2.
+func (r *registryType) Allocate(brickpath string) int {
+	r.Lock()
+	defer r.Unlock()
+
+	var port int
+	for p := r.basePort; p <= gfPortMax; p++ {
+		if r.ports[p].State == GfPmapPortFree || r.ports[p].State == GfPmapPortForeign {
+			// check if a port is free holding a lock on the port's "file"
+			if r.TryPortLock(p) {
+				if isPortFree(p) {
+					r.ports[p].State = GfPmapPortLeased
+					r.bricks[brickpath] = &brickStatus{
+						Port:     p,
+						SignedIn: false,
+					}
+					port = p
+					// keep port file locked if port was
+					// found to be free and we leased it to
+					// a brick
+					break
+				} else {
+					r.ports[p].State = GfPmapPortForeign
+					r.PortUnlock(p)
+				}
+			}
+		}
+	}
+
+	if port == 0 {
+		panic("registryAlloc() we ran out of free ports!")
+	}
+
+	if port > r.lastAlloc {
+		r.lastAlloc = port
+	}
+
+	return port
+}
+
+// Update marks the port used by a brick with a specified state. This
+// is called when a brick signs in.
+func (r *registryType) Update(port int, brickpath string, conn net.Conn) error {
+
+	if port < 0 || port > gfPortMax {
+		return fmt.Errorf("registry.Add(): invalid port %d", port)
+	}
+
+	r.Lock()
+	defer r.Unlock()
+
+	r.ports[port].State = GfPmapPortInUse
+	r.ports[port].Conn = conn
+	r.bricks[brickpath] = &brickStatus{
+		Port:     port,
+		SignedIn: true,
+	}
+
+	if r.lastAlloc < port {
+		r.lastAlloc = port
+	}
+
+	return nil
+}
+
+// SearchByBrickPath returns the port number used by the brick specified
+// by the brick path provided.
+func (r *registryType) SearchByBrickPath(brickpath string) (int, error) {
+
+	if brickpath == "" {
+		return -1, fmt.Errorf("SearchByBrickPath: brick path cannot be empty")
+	}
+
+	r.RLock()
+	defer r.RUnlock()
+
+	if b, ok := r.bricks[brickpath]; ok {
+		return b.Port, nil
+	}
+
+	return -1, fmt.Errorf("SearchByBrickPath: port for brick %s not found", brickpath)
+}
+
+// RemoveByConn deletes port map entry by brick process's TCP connection.
+// There will be only one TCP connection per brick process, regardless of
+// number of bricks in the process.
+func (r *registryType) RemovePortByConn(conn net.Conn) {
+
+	if conn == nil {
+		panic("RemovePortByConn(): conn passed is nil")
```

Comment:
Again, better to return an error here, than panic.

Reply:
Can do :+1: However, the call stack upwards has no provision to return the error anywhere (to user or brick). Will have the caller log it though.

Reply:
Done